File: sdks/runtime/python/lenny_runtime/tool.py
# SPDX-License-Identifier: MIT

"""§15.4.1 adapter-local tool surface.

An adapter-local tool call emits a stdout ``tool_call`` frame and blocks
until the matching ``tool_result`` frame arrives on stdin, correlating
by id. Unlike the platform MCP tools (which require Standard level),
adapter-local tools (read_file, write_file, list_dir, delete_file) are
resolved inside the adapter process with no MCP server.
"""

from __future__ import annotations

import secrets
import threading
from typing import Any

from .transport import FrameWriter
from .types import MessagePart, ToolResult


class _Pending:
    """In-flight bookkeeping for one ``tool_call``: an event that fires
    when the correlated ``tool_result`` arrives, plus the result slot
    and a rejection flag."""

    def __init__(self) -> None:
        self.event = threading.Event()
        self.result: ToolResult | None = None
        self.error: Exception | None = None


class ToolCallRegistry:
    """Correlates outbound §15.4.1 ``tool_call`` frames with inbound
    ``tool_result`` frames.

    The §15.4.1 frame loop delivers a ``tool_result`` here; an
    :class:`AdapterToolset` call registers and waits for one.
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._pending: dict[str, _Pending] = {}

    def register(self, call_id: str) -> _Pending:
        """Record a pending ``tool_call`` id and return its
        bookkeeping."""
        pending = _Pending()
        with self._lock:
            self._pending[call_id] = pending
        return pending

    def deliver(self, raw: dict[str, Any]) -> bool:
        """Route an inbound ``tool_result`` to the call that emitted the
        matching id. Reports whether a pending call was found."""
        call_id = str(raw.get("id", ""))
        with self._lock:
            pending = self._pending.pop(call_id, None)
        if pending is None:
            return False
        pending.result = ToolResult(
            content=[MessagePart.from_wire(p) for p in raw.get("content", [])],
            is_error=bool(raw.get("isError", False)),
        )
        pending.event.set()
        return True

    def cancel(self, call_id: str, err: Exception) -> None:
        """Drop a pending ``tool_call`` registration and fail its
        waiter."""
        with self._lock:
            pending = self._pending.pop(call_id, None)
        if pending is not None:
            pending.error = err
            pending.event.set()

    def reject_all(self, err: Exception) -> None:
        """Fail every pending call.

        The §15.4.1 frame loop calls it when the inbound stream closes
        so no caller waits forever.
        """
        with self._lock:
            pending_items = list(self._pending.items())
            self._pending.clear()
        for _, pending in pending_items:
            pending.error = err
            pending.event.set()


def _new_call_id() -> str:
    """Generate a unique §15.4.1 ``tool_call`` id with the recommended
    ``tc_`` prefix."""
    return "tc_" + secrets.token_hex(8)


def _tool_error(result: ToolResult, tool: str) -> None:
    """Raise when an error-flagged tool result is present.

    The adapter sets ``content[0].inline`` to the failure string (for
    example ``path_outside_workspace``).
    """
    if not result.is_error:
        return
    msg = "tool reported an error"
    if result.content and result.content[0].inline:
        msg = result.content[0].inline
    raise RuntimeError(f"{tool}: {msg}")


class AdapterToolset:
    """§15.4.1 adapter-local tool surface available at every integration
    level.

    It emits a stdout ``tool_call`` frame and blocks until the matching
    ``tool_result`` frame arrives on stdin.
    """

    def __init__(
        self,
        writer: FrameWriter,
        registry: ToolCallRegistry,
        timeout_s: float,
        slot_id: str | None,
    ) -> None:
        self._writer = writer
        self._registry = registry
        self._timeout_s = timeout_s if timeout_s > 0 else 30.0
        self._slot_id = slot_id

    def tool_call(self, name: str, arguments: dict[str, Any]) -> ToolResult:
        """Emit a §15.4.1 ``tool_call`` frame for the named
        adapter-local tool and block until the correlated
        ``tool_result`` arrives.

        The id is generated and unique within the process.
        """
        call_id = _new_call_id()
        pending = self._registry.register(call_id)
        frame: dict[str, Any] = {
            "type": "tool_call",
            "id": call_id,
            "name": name,
            "arguments": arguments,
        }
        if self._slot_id:
            frame["slotId"] = self._slot_id
        try:
            self._writer.write(frame)
        except Exception as err:
            self._registry.cancel(call_id, err)
            raise RuntimeError(f"write tool_call {name!r}: {err}") from err

        if not pending.event.wait(self._timeout_s):
            self._registry.cancel(
                call_id, TimeoutError(f"tool_call {name!r} timed out")
            )
            raise TimeoutError(
                f"tool_call {name!r} timed out after {self._timeout_s}s"
            )
        if pending.error is not None:
            raise pending.error
        assert pending.result is not None
        return pending.result

    def read_file(self, path: str) -> str:
        """Invoke the §15.4.1 ``read_file`` adapter-local tool.

        The path is confined to the pod workspace by the adapter; a path
        resolving outside /workspace returns an error result.
        """
        result = self.tool_call("read_file", {"path": path})
        _tool_error(result, "read_file")
        if not result.content:
            return ""
        return result.content[0].inline or ""

    def write_file(self, path: str, content: str) -> None:
        """Invoke the §15.4.1 ``write_file`` adapter-local tool, creating
        or overwriting a workspace file with UTF-8 content."""
        result = self.tool_call("write_file", {"path": path, "content": content})
        _tool_error(result, "write_file")

    def list_dir(self, path: str) -> list[MessagePart]:
        """Invoke the §15.4.1 ``list_dir`` adapter-local tool and return
        the directory entries the adapter reports."""
        result = self.tool_call("list_dir", {"path": path})
        _tool_error(result, "list_dir")
        return result.content

    def delete_file(self, path: str) -> None:
        """Invoke the §15.4.1 ``delete_file`` adapter-local tool."""
        result = self.tool_call("delete_file", {"path": path})
        _tool_error(result, "delete_file")

File: sdks/runtime/typescript/src/tool.ts
// SPDX-License-Identifier: MIT

// This module implements the §15.4.1 adapter-local tool surface. An
// adapter-local tool call emits a stdout tool_call frame and resolves
// when the matching tool_result frame arrives on stdin, correlating by
// id. Unlike the platform MCP tools (which require Standard level),
// adapter-local tools (read_file, write_file, list_dir, delete_file)
// are resolved inside the adapter process with no MCP server.

import { randomBytes } from "node:crypto";
import type { FrameWriter } from "./transport.js";
import type { AdapterTools, MessagePart, ToolResult } from "./types.js";

// InboundToolResult is the decoded §15.4.1 tool_result frame.
export interface InboundToolResult {
  type: string;
  id: string;
  content?: MessagePart[];
  isError?: boolean;
  slotId?: string;
}

// PendingToolCall is the in-flight bookkeeping for one tool_call: the
// callbacks that resolve when the correlated tool_result arrives.
interface PendingToolCall {
  resolve(result: InboundToolResult): void;
  reject(err: Error): void;
}

// ToolCallRegistry correlates outbound §15.4.1 tool_call frames with
// inbound tool_result frames. The §15.4.1 frame loop delivers a
// tool_result here; an AdapterTools call registers and awaits one.
export class ToolCallRegistry {
  private readonly pending = new Map<string, PendingToolCall>();

  // register records a pending tool_call id and returns a promise that
  // resolves with the correlated tool_result.
  register(id: string): Promise<InboundToolResult> {
    return new Promise<InboundToolResult>((resolve, reject) => {
      this.pending.set(id, { resolve, reject });
    });
  }

  // deliver routes an inbound tool_result to the call that emitted the
  // matching id. It reports whether a pending call was found.
  deliver(result: InboundToolResult): boolean {
    const entry = this.pending.get(result.id);
    if (!entry) {
      return false;
    }
    this.pending.delete(result.id);
    entry.resolve(result);
    return true;
  }

  // cancel drops a pending tool_call registration and rejects its
  // promise.
  cancel(id: string, err: Error): void {
    const entry = this.pending.get(id);
    if (entry) {
      this.pending.delete(id);
      entry.reject(err);
    }
  }

  // rejectAll fails every pending call. The §15.4.1 frame loop calls it
  // when the inbound stream closes so no caller waits forever.
  rejectAll(err: Error): void {
    for (const [id, entry] of this.pending) {
      this.pending.delete(id);
      entry.reject(err);
    }
  }
}

// newCallId generates a unique §15.4.1 tool_call id with the
// recommended tc_ prefix.
function newCallId(): string {
  return "tc_" + randomBytes(8).toString("hex");
}

// firstInline returns the inline content of the first result part, or
// throws when the tool reported a failure.
function firstInline(tr: ToolResult): string {
  toolError(tr, "tool");
  if (tr.content.length === 0) {
    return "";
  }
  return tr.content[0].inline ?? "";
}

// toolError throws when an error-flagged tool result is present. The
// adapter sets content[0].inline to the failure string (for example
// path_outside_workspace).
function toolError(tr: ToolResult, tool: string): void {
  if (!tr.isError) {
    return;
  }
  let msg = "tool reported an error";
  if (tr.content.length > 0 && tr.content[0].inline) {
    msg = tr.content[0].inline as string;
  }
  throw new Error(`${tool}: ${msg}`);
}

// AdapterToolset is the §15.4.1 adapter-local tool surface available at
// every integration level. It emits a stdout tool_call frame and
// resolves once the matching tool_result frame arrives on stdin.
export class AdapterToolset implements AdapterTools {
  constructor(
    private readonly writer: FrameWriter,
    private readonly registry: ToolCallRegistry,
    private readonly timeoutMs: number,
    private readonly slotId: string | undefined,
  ) {}

  // toolCall emits a §15.4.1 tool_call frame for the named
  // adapter-local tool and resolves once the correlated tool_result
  // arrives. The id is generated and unique within the process.
  async toolCall(
    name: string,
    args: Record<string, unknown>,
  ): Promise<ToolResult> {
    const id = newCallId();
    const wait = this.registry.register(id);
    const frame: Record<string, unknown> = {
      type: "tool_call",
      id,
      name,
      arguments: args,
    };
    if (this.slotId) {
      frame.slotId = this.slotId;
    }
    try {
      await this.writer.write(frame);
    } catch (err) {
      this.registry.cancel(id, err as Error);
      throw new Error(`write tool_call "${name}": ${(err as Error).message}`);
    }

    const timeout = this.timeoutMs > 0 ? this.timeoutMs : 30_000;
    let timer: NodeJS.Timeout | undefined;
    const guard = new Promise<never>((_, reject) => {
      timer = setTimeout(() => {
        const err = new Error(
          `tool_call "${name}" timed out after ${timeout}ms`,
        );
        this.registry.cancel(id, err);
        reject(err);
      }, timeout);
    });
    try {
      const result = await Promise.race([wait, guard]);
      return {
        content: result.content ?? [],
        isError: result.isError ?? false,
      };
    } finally {
      if (timer) {
        clearTimeout(timer);
      }
    }
  }

  // readFile invokes the §15.4.1 read_file adapter-local tool. The path
  // is confined to the pod workspace by the adapter; a path resolving
  // outside /workspace returns an error result.
  async readFile(path: string): Promise<string> {
    const tr = await this.toolCall("read_file", { path });
    return firstInline(tr);
  }

  // writeFile invokes the §15.4.1 write_file adapter-local tool,
  // creating or overwriting a workspace file with UTF-8 content.
  async writeFile(path: string, content: string): Promise<void> {
    const tr = await this.toolCall("write_file", { path, content });
    toolError(tr, "write_file");
  }

  // listDir invokes the §15.4.1 list_dir adapter-local tool and returns
  // the directory entries the adapter reports.
  async listDir(path: string): Promise<MessagePart[]> {
    const tr = await this.toolCall("list_dir", { path });
    toolError(tr, "list_dir");
    return tr.content;
  }

  // deleteFile invokes the §15.4.1 delete_file adapter-local tool.
  async deleteFile(path: string): Promise<void> {
    const tr = await this.toolCall("delete_file", { path });
    toolError(tr, "delete_file");
  }
}
